stablehlo/dialect/VhloOps.cpp

```diff
@@ -16,12 +16,140 @@ limitations under the License.
+  return conv.convertType(type);
+}
+
+Type convertTypeToVhloForParse(Type type) {
+  struct BuiltinToVhloParseConverter : VhloTypeConverter {
+    BuiltinToVhloParseConverter() : VhloTypeConverter() {
+      addBuiltinToVhloConversions();
+    }
+    Attribute convertEncoding(Attribute attr) override { return attr; }
+  };
+  BuiltinToVhloParseConverter conv;
+  return conv.convertType(type);
+}
+
+}  // namespace
+// Helper functions for VHLO printers and parsers
+static void printEncoding(AsmPrinter& os, Attribute encoding) {
+  if (!encoding) return;
+  os << ", " << encoding;
+}
+
+ParseResult parseEncoding(AsmParser& parser, Attribute& encoding) {
+  if (failed(parser.parseOptionalComma())) {
+    return success();
+  }
+  if (failed(parser.parseAttribute(encoding))) return failure();
+  return success();
+}
+
+static void printShape(AsmPrinter& os, ArrayRef<int64_t> dimSizes) {
+  if (dimSizes.empty()) return;
+  for (int64_t dimSize : dimSizes) {
+    os << hlo::dimSizeToString(dimSize) << 'x';
+  }
+}
+
+ParseResult parseShape(AsmParser& parser, SmallVector<int64_t>& dimSizes) {
+  if (failed(parser.parseDimensionList(dimSizes))) {
+    return failure();
+  }
+  return success();
+}
+
+static void printAttributeArray(AsmPrinter& os, ArrayRef<Attribute> arrayAttr) {
+  os << '[' << arrayAttr << ']';
+}
+
+ParseResult parseAttributeArray(AsmParser& parser,
+                                SmallVector<Attribute>& arrayAttr) {
+  ArrayAttr array;
+  if (failed(parser.parseAttribute(array))) {
+    return failure();
+  }
+  arrayAttr.append(array.begin(), array.end());
+  return success();
+}
+
+void IntegerV1Attr::print(mlir::AsmPrinter& p) const {
+  p << '<'
+    << IntegerAttr::get(convertTypeToBuiltinForPrint(getType()), getValue())
+    << '>';
+}
+
+Attribute IntegerV1Attr::parse(AsmParser& parser, mlir::Type) {
+  IntegerAttr attr;
+  if (failed(parser.parseLess()) || failed(parser.parseAttribute(attr)) ||
+      failed(parser.parseGreater())) {
+    return IntegerV1Attr();
+  }
+  return IntegerV1Attr::get(parser.getContext(),
+                            convertTypeToVhloForParse(attr.getType()),
+                            attr.getValue());
+}
+
+void DenseIntOrFPElementsV1Attr::print(mlir::AsmPrinter& p) const {
+  p << '<'
+    << DenseIntOrFPElementsAttr::getFromRawBuffer(
+           convertTypeToBuiltinForPrint(getType()), getRawData())
+    << '>';
+}
+
+Attribute DenseIntOrFPElementsV1Attr::parse(AsmParser& parser, mlir::Type) {
+  DenseIntOrFPElementsAttr attr;
+  if (failed(parser.parseLess()) || failed(parser.parseAttribute(attr)) ||
+      failed(parser.parseGreater())) {
+    return DenseIntOrFPElementsV1Attr();
+  }
+  return DenseIntOrFPElementsV1Attr::get(
+      parser.getContext(), convertTypeToVhloForParse(attr.getType()),
+      attr.getRawData());
+}
+
+}  // namespace vhlo
+}  // namespace mlir
+
 // Include order matters
 #include "stablehlo/dialect/VhloTypeInterfaces.cpp.inc"
 #define GET_TYPEDEF_CLASSES
```

Comment:
Should we move these into VhloTypes.cpp as well?

Reply:
Let's think about that later. I added a note to the Vhlo sync to spec issue. This cpp is required by things later in this file.

Reply:
I think this is important because keeping these `.cpp.inc` files here will make the `vhlo_types` target non self-contained. To address the dependency from this file to these definitions, can we include `VhloTypes.h` from here?

Reply:
We agreed to fix this early next week.